jetty9-compat-base/src/main/java/com/google/apphosting/runtime/jetty9/NoOpSession.java

```diff
@@ -0,0 +1,191 @@
+   * @param context the ServletContext to which the session belongs
+   */
+  protected NoOpSession(String id, String nodeId, ServletContext context) {
+    this.clusterId = id;
+    this.nodeId = nodeId;
+    this.context = context;
+  }
+
+  /**
+   * Finish a request
+   */
+  protected void complete() {
+    attributes.clear();
+  }
+
+  @Override
+  public long getCreationTime() {
+    checkValid();
+    return 0;
+  }
+
+  @Override
+  public String getId() {
+    return clusterId;
+  }
+
+  /**
+   * @return the session id with worker suffix
+   */
+  protected String getNodeId() {
+    return nodeId;
+  }
+
+  @Override
+  public long getLastAccessedTime() {
+    checkValid();
+    return 0;
+  }
+
+  @Override
+  public ServletContext getServletContext() {
+    return context;
+  }
+
+  @Override
+  public void setMaxInactiveInterval(int interval) {}
+
+  @Override
+  public int getMaxInactiveInterval() {
+    return 0;
+  }
+
+  @Deprecated
+  @Override
+  public HttpSessionContext getSessionContext() {
+    checkValid();
+    return null;
+  }
+
+  @Override
+  public Object getAttribute(String name) {
+    synchronized (this) {
+      checkValid();
+      return attributes.get(name);
+    }
+  }
+
+  @Deprecated
+  @Override
+  public Object getValue(String name) {
+    return getAttribute(name);
+  }
+
+  @Override
+  public Enumeration<String> getAttributeNames() {
+    synchronized (this) {
+      checkValid();
+      List<String> names = new ArrayList<String>(attributes.keySet());
+      return Collections.enumeration(names);
+    }
+  }
+
+  @Deprecated
+  @Override
+  public String[] getValueNames() {
+    synchronized (this) {
+      checkValid();
+      List<String> names = new ArrayList<String>(attributes.keySet());
+      if (names.isEmpty()) {
+        return new String[0];
+      }
+      return names.toArray(new String[names.size()]);
+    }
+  }
+
+  @Override
+  public void setAttribute(String name, Object value) {
+    synchronized (this) {
+      attributes.put(name, value);
+    }
+  }
+
+  @Deprecated
+  @Override
+  public void putValue(String name, Object value) {
+    setAttribute(name, value);
+  }
+
+  @Override
+  public void removeAttribute(String name) {
+    synchronized (this) {
+      attributes.remove(name);
+    }
+  }
+
+  @Deprecated
+  @Override
+  public void removeValue(String name) {
+    removeAttribute(name);
+  }
+
+  @Override
+  public void invalidate() {
+    synchronized (this) {
+      checkValid();
+      valid = false;
+      attributes.clear();
+    }
+  }
+
+  /**
+   * @return true if the session has not been invalidated
+   */
+  protected boolean isValid() {
+    return valid;
+  }
+
+  @Override
+  public boolean isNew() {
+    checkValid();
+    return true;
+  }
+
+  private void checkValid() throws IllegalStateException {
+    if (!isValid()) throw new IllegalStateException("Session " + clusterId + " invalid");
```

Comment:
Any reason for not using Guava's Preconditions? could be one-liner.
If you prefer as is then change it to:
```java
if (!isValid()) {
  throw new IllegalStateException("Session " + clusterId + " invalid")
}
```

otherwise you could do:
```java
Preconditions.checkState(isValid(), "Session %s invalid", clusterId);
```